plugins/s3/src/plugin.cpp

```diff
@@ -0,0 +1,277 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <tenzir/argument_parser.hpp>
+#include <tenzir/location.hpp>
+#include <tenzir/plugin.hpp>
+
+#include <arrow/filesystem/filesystem.h>
+#include <arrow/filesystem/s3fs.h>
+#include <arrow/filesystem/type_fwd.h>
+#include <arrow/io/api.h>
+#include <arrow/util/uri.h>
+#include <fmt/core.h>
+
+namespace tenzir::plugins::s3 {
+
+namespace {
+
+struct s3_args {
+  bool anonymous;
+  located<std::string> uri;
+
+  template <class Inspector>
+  friend auto inspect(Inspector& f, s3_args& x) -> bool {
+    return f.object(x).pretty_name("s3_args").fields(
+      f.field("anonymous", x.anonymous), f.field("uri", x.uri));
+  }
+};
+
+// We use 2^20 for the upper bound of a chunk size, which exactly matches the
+// upper limit defined by execution nodes for transporting events.
+// TODO: Get the backpressure-adjusted value at runtime from the execution node.
+constexpr size_t max_chunk_size = 1 << 20;
+} // namespace
+
+class s3_loader final : public plugin_loader {
+public:
+  s3_loader() = default;
+
+  s3_loader(s3_args args) : args_{std::move(args)} {
+  }
+  auto instantiate(operator_control_plane& ctrl) const
+    -> std::optional<generator<chunk_ptr>> override {
+    return
+      [](s3_args args, operator_control_plane& ctrl) -> generator<chunk_ptr> {
+        auto initialized = arrow::fs::EnsureS3Initialized();
+        if (not initialized.ok()) {
+          diagnostic::error("failed to initialize Arrow S3 functionality: {}",
+                            initialized.ToString())
+            .emit(ctrl.diagnostics());
+          co_return;
+        }
+        auto guard = caf::detail::make_scope_guard([&ctrl]() {
+          auto finalized = arrow::fs::FinalizeS3();
+          if (not finalized.ok()) {
+            diagnostic::error("failed to close Arrow S3 filesystem: {}",
+                              finalized.ToString())
+              .emit(ctrl.diagnostics());
+          }
+        });
+        auto fs = arrow::fs::S3FileSystem::Make(
+          (args.anonymous) ? arrow::fs::S3Options::Anonymous()
+                           : arrow::fs::S3Options::Defaults());
```

Comment:
Is it maybe possible to take a full URI here?
This would allow the user to change the region, scheme and endpoint.

I have following working:
```
auto fs = arrow::fs::FileSystemFromUri(args.uri.inner, &path);
```

Then for most of the rest of the file using `path` instead of `args.uri.inner`.

The pipeline usage then looks like for S3-compatible storage:
```
from s3 s3://my_bucket/my_file?endpoint_override=s3.us-west-002.backblazeb2.com
```

For S3:
```
from s3 s3://my_bucket/my_file
```

Inspiration taken from https://github.com/apache/arrow/blob/cf0cf9bdaaad89d603fe0d6b1daa8a482f3670b7/cpp/src/arrow/filesystem/s3fs_test.cc#L292C12-L292C88

Reply:
It looks possible, but `FileSystemFromUri` would also create `Arrow::Filesystem`s for other URIs (e.g. `file://` or `gcs://`), so I would have to implement an additional `arrow::fs::S3FileSystem` check. 

But yeah, it is worth a try. Let me try.

Reply:
BTW: Your example is showing `S3Options::FromUri` - since the `S3Options` is the one struct that holds credential info (with overrides), we may not even need a `FileSystemFromUri` call - we can just call `S3FileSystem::Make` with the options from `S3Options::FromUri`. But I would need to try that first.

OK?

Reply:
Yeah, it's actually the one I tried first but `S3Options::FromUri` returns a `Result<S3Options>`. Before I could find how to convert that into just a `S3Options` (didn't search further), I found `FileSystemFromUri` and got the `FileSystemFromUri` version to work.

Reply:
That's alright, I found a way to convert Arrow's fancy `Result`s.
Now I just need to test around a bit.

Reply:
I refreshed the PR with an implementation that uses `S3Options::FromUri` with `S3Options::Defaults()` as a fallback.

Reply:
Thanks, I've verified that it does work for non-AWS endpoints too :-)

Reply:
Thank you for your support! :pray: